### PR TITLE
Refactoring in client.py/transport.py

### PR DESCRIPTION
**Pull request by fredo**
## Description
With the recent changes, some small refactorings and optimizations are possible to make the code more readable. 

This includes:

- remove the redundancy of `should_listen` and `stop_event`
- remove `first_sync` flag at sync
- prevent possible attack vector of joining rooms with large number of addresses due to ecrecover

Fixes: #5921

### CODE REVIEW

File: raiden/network/transport/matrix/client.py [hunk 1]
```diff
@@ -441,8 +441,7 @@ def start_listener_thread(
             exception_handler: Optional exception handler function which can
                 be used to handle exceptions in the caller thread.
         """
-        assert not self.should_listen and self.sync_worker is None, "Already running"
-        self.should_listen = True
+        assert self.sync_worker is None, "Already running"
```
Comment (hackaugusto) [inline raiden/network/transport/matrix/client.py]: shouldn't we have an assert for the state of the `stop_event`? Because `Runnable`s have the concept of `start` and `stop`, it could happen the even is not properly reset.

File: raiden/network/transport/matrix/client.py [hunk 2]
```diff
@@ -696,13 +693,6 @@ def _handle_message(
         while True:
             gevent.wait({response_queue, stop_event}, count=1)
 
-            if stop_event.is_set():
```
Comment (hackaugusto) [inline raiden/network/transport/matrix/client.py]: Hummm, why did you move this code bellow?

Comment (fredo) [reply raiden/network/transport/matrix/client.py]: I moved it there so that we dont run into the case where we have a lot of responses in the queue and handle all of them before getting back to the condition. This was caught before by should listen. The `if` in that position replaces it and renders the old place (where it was) useless.

File: raiden/network/transport/matrix/client.py (hunk 1 above)
Comment (fredo) [reply raiden/network/transport/matrix/client.py]: sounds reasonable. What is your proposed solution? 

`assert self.stop_event.is_set() and self.sync_worker is None, "Already running"`

Comment (fredo) [reply raiden/network/transport/matrix/client.py]: although the client is not a runnable right?

Comment (hackaugusto) [reply raiden/network/transport/matrix/client.py]: the client not, but the transport yes, I believe the stop event is shared, no?

Comment (fredo) [reply raiden/network/transport/matrix/client.py]: AFAIK its only used within the client
